File: src/chats/src/updates/remove_participant.rs
use ic_cdk::export::candid::CandidType;
use ic_cdk::storage;
use shared::user_id::UserId;
use crate::domain::chat::{ChatId, ChatEnum};
use crate::domain::chat_list::ChatList;

pub fn update(chat_id: ChatId, user: UserId) -> Result {
    let chat_list: &mut ChatList = storage::get_mut();
    let me = shared::user_id::get_current();

    let chat = chat_list.get_mut(chat_id, &me);

    match chat {
        Some(ChatEnum::Group(group_chat)) => {
            if me == user {
                return Result::CannotRemoveSelfFromChat;
            }

            match group_chat.remove_participant(&me, &user) {
                Some(true) => Result::Success,
                Some(false) => Result::ParticipantNotFound,
                None => Result::Unauthorized
            }
        },
        Some(_) => Result::NotGroupChat,
        None => Result::ChatNotFound
    }
}

#[derive(CandidType)]
pub enum Result {
    Success,
    Unauthorized,
    ParticipantNotFound,
    CannotRemoveSelfFromChat,
    ChatNotFound,
    NotGroupChat
}

File: src/website/services/chats/removeParticipant.ts
import { UserId } from "../../domain/model/users";
import { ChatId } from "../../domain/model/chats";
import { toCandid as userIdToCandid } from "../candidConverters/userId";
import CanisterClientFactory from "../CanisterClientFactory";

export default async function(chatId: ChatId, userId: UserId) : Promise<RemoveParticipantResponse> {
    const client = CanisterClientFactory.current!.chatsClient;
    const response = await client.remove_participant(chatId, userIdToCandid(userId));

    if ("Success" in response) {
        return RemoveParticipantResponse.Success;
    } else if ("Unauthorized" in response) {
        return RemoveParticipantResponse.Unauthorized;
    } else if ("ParticipantNotFound" in response) {
        return RemoveParticipantResponse.ParticipantNotFound;
    } else if ("CannotRemoveSelfFromChat" in response) {
        return RemoveParticipantResponse.CannotRemoveSelfFromChat;
    } else if ("ChatNotFound" in response) {
        return RemoveParticipantResponse.ChatNotFound;
    } else if ("NotGroupChat" in response) {
        return RemoveParticipantResponse.NotGroupChat;
    } else {
        throw new Error("Unrecognised 'remove_participant' response");
    }
}

export enum RemoveParticipantResponse {
    Success,
    Unauthorized,
    ParticipantNotFound,
    CannotRemoveSelfFromChat,
    ChatNotFound,
    NotGroupChat
}